File: feishu_entry.py
import json
import os
import queue
import time
import threading

import lark_oapi as lark
from lark_oapi.api.im.v1 import (
    CreateMessageRequest,
    CreateMessageRequestBody,
    P2ImMessageReceiveV1,
    ReplyMessageRequest,
    ReplyMessageRequestBody,
)

from channel_layer import normalize_feishu_event


class FeishuEntry:
    """Channel 适配层：收飞书消息、去重、标准化、入队、发回复。"""

    def __init__(self, runtime):
        self.runtime = runtime
        self.app_id = os.getenv("LARK_APP_ID", "").strip()
        self.app_secret = os.getenv("LARK_APP_SECRET", "").strip()
        if not self.app_id or not self.app_secret:
            raise RuntimeError("Missing LARK_APP_ID/LARK_APP_SECRET for feishu message entry.")

        self.client = lark.Client.builder().app_id(self.app_id).app_secret(self.app_secret).build()
        self._dedup_ttl_seconds = int(os.getenv("FEISHU_DEDUP_TTL_SECONDS", "7200"))
        self._recent_event_keys = {}
        self._dedup_lock = threading.Lock()

        # 队列只负责串行处理消息：上一条完成后再处理下一条。
        self._message_queue = queue.Queue()
        self._worker_thread = threading.Thread(target=self._worker_loop, daemon=True)
        self._worker_thread.start()

        self.event_handler = (
            lark.EventDispatcherHandler.builder("", "")
            .register_p2_im_message_receive_v1(self._on_message)
            .build()
        )
        self.ws_client = lark.ws.Client(
            self.app_id,
            self.app_secret,
            event_handler=self.event_handler,
            log_level=lark.LogLevel.DEBUG,
        )

    def send_reply(self, data: P2ImMessageReceiveV1, text: str) -> None:
        content = json.dumps({"text": text})
        message = data.event.message

        if message.chat_type == "p2p":
            request = (
                CreateMessageRequest.builder()
                .receive_id_type("chat_id")
                .request_body(
                    CreateMessageRequestBody.builder()
                    .receive_id(message.chat_id)
                    .msg_type("text")
                    .content(content)
                    .build()
                )
                .build()
            )
            resp = self.client.im.v1.message.create(request)
            if not resp.success():
                raise RuntimeError(f"send p2p failed: {resp.code}, {resp.msg}, {resp.get_log_id()}")
            return

        request = (
            ReplyMessageRequest.builder()
            .message_id(message.message_id)
            .request_body(
                ReplyMessageRequestBody.builder().content(content).msg_type("text").build()
            )
            .build()
        )
        resp = self.client.im.v1.message.reply(request)
        if not resp.success():
            raise RuntimeError(f"reply group failed: {resp.code}, {resp.msg}, {resp.get_log_id()}")

    def _on_message(self, data: P2ImMessageReceiveV1) -> None:
        msg = normalize_feishu_event(data)
        dedup_key = self._build_dedup_key(data, msg.message_id, msg.session_id, msg.text)

        if self._is_duplicate_event(dedup_key):
            print(f"[DEBUG] skip duplicated feishu event: {dedup_key}")
            return

        if not msg.text:
            self._message_queue.put(("reply_text", data, "parse message failed, please send text message"))
            return

        self._message_queue.put(("handle_runtime", data, msg))

    def _worker_loop(self) -> None:
        while True:
            task_type, data, content = self._message_queue.get()
            try:
                if task_type == "reply_text":
                    self.send_reply(data, content)
                elif task_type == "handle_runtime":
                    self.runtime.process_channel_message(self, data, content)
            except Exception as e:
                print(f"[ERROR] worker handle message failed: {e}")
                try:
                    self.send_reply(data, "处理消息时发生错误，请稍后重试。")
                except Exception as send_err:
                    print(f"[ERROR] send error message failed: {send_err}")
            finally:
                self._message_queue.task_done()

    def _build_dedup_key(self, data: P2ImMessageReceiveV1, message_id: str, session_id: str, text: str) -> str:
        event_id = ""
        try:
            event_id = (data.header.event_id or "").strip()
        except Exception:
            event_id = ""
        if event_id:
            return f"event_id:{event_id}"
        if message_id:
            return f"message_id:{message_id}"
        return f"session_text:{session_id}:{text}"

    def _is_duplicate_event(self, dedup_key: str) -> bool:
        now = time.time()
        with self._dedup_lock:
            expired_keys = [
                key for key, expires_at in self._recent_event_keys.items()
                if expires_at <= now
            ]
            for key in expired_keys:
                self._recent_event_keys.pop(key, None)

            if dedup_key in self._recent_event_keys:
                return True
            self._recent_event_keys[dedup_key] = now + self._dedup_ttl_seconds
            return False

    def run(self) -> None:
        self.ws_client.start()

File: src/feishu-entry.ts
import type { AgentRuntime } from "./agent-runtime.js";
import type { RuntimeMessage, RuntimeResult } from "./types.js";
import { normalizeFeishuEvent } from "./channel-layer.js";

interface QueueTask {
  type: "reply_text" | "handle_runtime";
  data: any;
  content: string | RuntimeMessage;
}

export class FeishuEntry {
  private readonly appId = (process.env.LARK_APP_ID ?? "").trim();
  private readonly appSecret = (process.env.LARK_APP_SECRET ?? "").trim();
  private readonly dedupTtlSeconds = Number(process.env.FEISHU_DEDUP_TTL_SECONDS ?? "7200");
  private readonly recentEventKeys = new Map<string, number>();
  private readonly activeConversations = new Map<string, string>();
  private queueChain: Promise<void> = Promise.resolve();
  private client: any;
  private wsClient: any;
  private eventDispatcher: any;
  private larkModule: any;

  constructor(private readonly runtime: AgentRuntime) {
    if (!this.appId || !this.appSecret) {
      throw new Error("Missing LARK_APP_ID/LARK_APP_SECRET for feishu message entry.");
    }
  }

  private newConversationId(): string {
    return `conv-${Math.random().toString(16).slice(2, 12)}`;
  }

  private conversationKey(userScopeId: string, chatId: string): string {
    return `${userScopeId}:${chatId}`;
  }

  private getOrCreateConversationId(userScopeId: string, chatId: string): string {
    const key = this.conversationKey(userScopeId, chatId);
    const existing = this.activeConversations.get(key);
    if (existing) {
      return existing;
    }
    const created = this.newConversationId();
    this.activeConversations.set(key, created);
    return created;
  }

  private setConversationId(userScopeId: string, chatId: string, conversationId: string): void {
    if (!conversationId) {
      return;
    }
    this.activeConversations.set(this.conversationKey(userScopeId, chatId), conversationId);
  }

  private async init(): Promise<void> {
    if (this.client && this.wsClient && this.eventDispatcher) {
      return;
    }

    const imported = await import("@larksuiteoapi/node-sdk");
    this.larkModule = (imported as any).default ?? imported;

    this.client = new this.larkModule.Client({ appId: this.appId, appSecret: this.appSecret });
    this.wsClient = new this.larkModule.WSClient({ appId: this.appId, appSecret: this.appSecret });

    if (this.larkModule.EventDispatcher) {
      this.eventDispatcher = new this.larkModule.EventDispatcher({}).register({
        "im.message.receive_v1": async (data: any) => this.onMessage(data),
      });
      return;
    }

    if (this.larkModule.EventDispatcherHandler?.builder) {
      this.eventDispatcher = this.larkModule.EventDispatcherHandler.builder("", "")
        .register_p2_im_message_receive_v1(async (data: any) => this.onMessage(data))
        .build();
      return;
    }

    throw new Error("Unsupported @larksuiteoapi/node-sdk event dispatcher API.");
  }

  async sendReply(data: any, text: string): Promise<void> {
    await this.init();
    const content = JSON.stringify({ text });
    const message = data?.event?.message ?? {};

    let resp: any;
    if (message.chat_type === "p2p") {
      resp = await this.client.im.message.create({
        params: { receive_id_type: "chat_id" },
        data: {
          receive_id: message.chat_id,
          msg_type: "text",
          content,
        },
      });
    } else {
      resp = await this.client.im.message.reply({
        path: { message_id: message.message_id },
        data: {
          msg_type: "text",
          content,
        },
      });
    }

    if (resp && typeof resp.code === "number" && resp.code !== 0) {
      throw new Error(`send feishu message failed: ${resp.code}, ${resp.msg ?? ""}`);
    }
  }

  private onMessage(data: any): void {
    const baseMsg = normalizeFeishuEvent(data);
    const conversationId = this.getOrCreateConversationId(baseMsg.user_scope_id, baseMsg.chat_id);
    const msg = normalizeFeishuEvent(data, conversationId);
    const dedupKey = this.buildDedupKey(data, msg.message_id, msg.session_id, msg.text);

    if (this.isDuplicateEvent(dedupKey)) {
      console.log(`[DEBUG] skip duplicated feishu event: ${dedupKey}`);
      return;
    }

    msg.dedup_key = dedupKey;
    msg.enqueue_ts_ms = Date.now();

    if (!msg.text) {
      this.enqueue({ type: "reply_text", data, content: "parse message failed, please send text message" });
      return;
    }

    this.enqueue({ type: "handle_runtime", data, content: msg });
  }

  private enqueue(task: QueueTask): void {
    this.queueChain = this.queueChain
      .then(() => this.handleTask(task))
      .catch((error) => {
        console.error(`[ERROR] worker handle message failed: ${error instanceof Error ? error.message : String(error)}`);
      });
  }

  private async handleTask(task: QueueTask): Promise<void> {
    try {
      if (task.type === "reply_text") {
        await this.sendReply(task.data, String(task.content));
        return;
      }

      const msg = task.content as RuntimeMessage;
      const queueWaitMs = msg.enqueue_ts_ms ? Math.max(0, Date.now() - msg.enqueue_ts_ms) : 0;
      console.log(
        `[DEBUG] queue_metrics session_id=${msg.session_id} conversation_id=${msg.conversation_id ?? ""} queue_wait_ms=${queueWaitMs} dedup_key=${msg.dedup_key ?? ""}`,
      );
      const result: RuntimeResult = await this.runtime.processChannelMessage(this, task.data, msg);
      this.setConversationId(msg.user_scope_id, msg.chat_id, result.conversation_id);
    } catch (error) {
      console.error(`[ERROR] worker handle message failed: ${error instanceof Error ? error.message : String(error)}`);
      try {
        await this.sendReply(task.data, "处理消息时发生错误，请稍后重试。");
      } catch (sendError) {
        console.error(`[ERROR] send error message failed: ${sendError instanceof Error ? sendError.message : String(sendError)}`);
      }
    }
  }

  private buildDedupKey(data: any, messageId: string, sessionId: string, text: string): string {
    const eventId = String(data?.header?.event_id ?? "").trim();
    if (eventId) {
      return `event_id:${eventId}`;
    }
    if (messageId) {
      return `message_id:${messageId}`;
    }
    return `session_text:${sessionId}:${text}`;
  }

  private isDuplicateEvent(dedupKey: string): boolean {
    const now = Date.now() / 1000;
    for (const [key, expiresAt] of this.recentEventKeys.entries()) {
      if (expiresAt <= now) {
        this.recentEventKeys.delete(key);
      }
    }

    if (this.recentEventKeys.has(dedupKey)) {
      return true;
    }
    this.recentEventKeys.set(dedupKey, now + this.dedupTtlSeconds);
    return false;
  }

  async run(): Promise<void> {
    await this.init();
    if (typeof this.wsClient.start !== "function") {
      throw new Error("Feishu WS client does not expose start().");
    }

    const startResult = this.wsClient.start({ eventDispatcher: this.eventDispatcher });
    if (startResult && typeof startResult.then === "function") {
      await startResult;
    }
  }
}
